.circleci/config.yml

```diff
@@ -1,79 +1,127 @@
+              --build-arg "GIT_SHA1=$CIRCLE_SHA1" \
+              --build-arg "LICENSE=GPL-3.0-only" \
+              --build-arg "VCS_REF=$CIRCLE_SHA1" \
+              --build-arg "VENDOR=Reaction Commerce" \
+              -t "$DOCKER_REPOSITORY:$CIRCLE_SHA1" .
+            mkdir -p docker-cache
+            docker save \
+              -o docker-cache/docker-image.tar \
+              "$DOCKER_REPOSITORY:$CIRCLE_SHA1"
+      - persist_to_workspace:
+          root: docker-cache
           paths:
-            - ~/docker-cache
-
-  test:
-    docker:
-      - image: docker:latest
+            - docker-image.tar
+            - docker-tags.txt
 
+  docker-push:
+    <<: *defaults
     steps:
       - setup_remote_docker
-      - checkout
-
-      - restore_cache:
-          key: docker-cache-{{ .Revision }}
-
+      - attach_workspace:
+          at: docker-cache
       - run:
           name: Load Docker Image
           command: |
-            docker load < ~/docker-cache/image.tar
-
+            docker load < docker-cache/docker-image.tar
       - run:
-          name: Test installed dependencies
+          name: Tag Docker Image
           command: |
-            docker run --rm reactioncommerce/base:latest bash test-script.sh
-
-  deploy:
-    docker:
-      - image: docker:latest
+            cat docker-cache/docker-tags.txt \
+              | xargs -t -I % \
+                docker tag \
+                  "$DOCKER_REPOSITORY:$CIRCLE_SHA1" \
+                  "$DOCKER_REPOSITORY:%"
+      - run:
+          name: Docker Push
+          command: |
+            docker login -u "$DOCKER_USER" -p "$DOCKER_PASS"
+            docker push "$DOCKER_REPOSITORY:$CIRCLE_SHA1"
+            cat docker-cache/docker-tags.txt \
+              | xargs -t -I % \
+                docker push "$DOCKER_REPOSITORY:%"
 
+  test:
+    <<: *defaults
     steps:
       - setup_remote_docker
-      - run: apk add --update bash git openssh
-      - checkout
-
-      - restore_cache:
-          key: docker-cache-{{ .Revision }}
-
-      - run: .circleci/deploy.sh
+      - attach_workspace:
+          at: docker-cache
+      - run:
+          name: Load Docker Image
+          command: |
+            docker load < docker-cache/docker-image.tar
+      - run:
+          name: Test installed dependencies
+          command: |
+            docker run --rm \
+              "$DOCKER_REPOSITORY:$CIRCLE_SHA1" \
+              bash test-script.sh
 
 workflows:
   version: 2
-  build_and_deploy:
+  build_and_test:
     jobs:
-      - build:
-          filters:
-            tags:
-              only: /.*/
-      - test:
-          requires:
-            - build
-          filters:
-            tags:
-              only: /.*/
-      - deploy:
+      - docker-build:
+          context: reaction-build-read
+      - docker-push:
+          context: reaction-publish-docker
           requires:
-            - build
+            - docker-build
             - test
```

Comment:
I see that the deploy step depends on test but I'd posit that we should always push the image. Our source of truth is git (Github) and the released artifacts should mirror that reality. Otherwise we can have confusing scenarious where we expect a Docker image to contain something and it doesn't. Fixes to tests would be forward commits in Github and thus would be forward published artifacts by the build process.

Reply:
👍